Use final URLs after redirects in menus

lukasbestle opened the pull request:
Some of our templates redirect to another page (e.g. child page). If we include those pages in menus, the user needs to be redirected even though we already know the final target. This is currently mostly the case with `docs/guide` and `license`.

This PR implements two new page methods:

- `menuUrl()`: Returns the final URL after redirects to be used in menus. Overridden in the page model where relevant, otherwise falls back to the page URL.
- `isActiveInMenu()`: Variant of `isActive()` that respects `menuUrl()`. This fixes a bug in the menu for legal pages where the [License page](https://getkirby.com/license) wouldn't show an active state because it immediately redirects to its last child.

lukasbestle committed:
Page methods for menu URLs

lukasbestle committed:
Use menu URL in all menus

lukasbestle committed:
Add overrides for menu URLs in relevant models

distantnative commented:
@lukasbestle Could you please share why we need to replace `->isOpen()` with `->isActiveInMenu()`? Based on the description above, I would this that `isOpen` is exactly made for that.

And how about if we overwrite the `url()` method for those instead of creating a new `menuUrl()` for all object? Do you see a case where those models still need their original `url()`?

lukasbestle commented:
@distantnative On `isOpen()`: That's basically the opposite of what we need. We explicitly don't want to highlight the main menu item if a subpage is active. We do use `isOpen` for controlling whether the menu is expanded by default, but the other use we had of `isOpen` for the submenu was wrong IMO. Inside a submenu, we are already displaying the page itself and there are no additional children. Or have I misunderstood the structure?

About replacing the `url()` method: That's what I had before, but unfortunately it won't work for this use case. If the parent returns the URL of its child, the child's `url()` method then causes an infinite loop because it uses its parent's `url()` as a base for adding its own UID at the end.

lukasbestle committed:
Revert to `isOpen` for submenus

lukasbestle commented:
Thinking about it again, I think you are right. We don't have recursive submenus, only two levels (main and one sub level). So in the sub level, the menu item should be highlighted even if one of its children is active. So `isOpen()` is correct there. And I can't think of a case when that would be broken by redirects.

distantnative commented on site/plugins/site/extensions/pageMethods.php:
```diff
@@ -0,0 +1,26 @@
+	 * fallback if the page model doesn't override it
+	 */
+	'menuUrl' => fn () => $this->url()
+];
```
`Suggestion:` We could use a default page model instead (https://getkirby.com/docs/guide/templates/page-models#default-page-model) - that way those would be together with the other page models in `site/models` which could make it easier to understand/keep an overview later down the road.

Would though include that the other page models have to extend the `DefaultPage` model then.

distantnative commented on site/models/buzz-entry.php:
```diff
@@ -18,6 +18,11 @@ public function isExternalLink(): bool
+	{
+		return $this->url();
+	}
+
```
```suggestion
```

`Todo:` This looks to be the default as in the custom page method (potential default page model). So I don't think we need to add this here.

distantnative commented on site/models/terms.php:
```diff
@@ -0,0 +1,24 @@
+
+	public function menuUrl(): string
+	{
+		return $this->latestVersion()->url() ?? $this->url();
```
```suggestion
		return $this->latestVersion()?->url() ?? $this->url();
```

distantnative commented on site/snippets/layouts/footer.php:
```diff
@@ -80,11 +80,11 @@
+							<li><a href="<?= page('privacy')->menuUrl() ?>">Privacy</a></li>
+							<li><a href="<?= page('license')->menuUrl() ?>">License</a></li>
+							<li><a href="<?= page('press')->menuUrl() ?>">Presskit</a></li>
+							<li><a href="<?= page('contact')->menuUrl() ?>">Contact</a></li>
```
`Suggestion:` To keep the footer and main menus DRY-er, we could use a snippet which we just pass something like

```php
[
    'For developers' => page('features/developers'),
    'For designers' => page('features/designers'),
    'Feedback' => 'https://feedback.getkirby.com'
]
```

And the snippet renders each `<li>` depending whether it contains a page object or a string (for external) differently).

lukasbestle committed:
Use a default page model

lukasbestle committed:
Override proper `isActive` method

lukasbestle committed:
Fix fallback on `null`

lukasbestle committed:
Fix highlighting in menu

URL must be absolute for comparing URLs in the menu (`isActive`)

lukasbestle committed:
Remove obsolete overrides

New menu handling works without them

lukasbestle committed:
Remove duplicated method

Implementation is the same as the default implementation

lukasbestle committed:
Fix menu URLs with relative links

Content files contain relative links, menu URLs need to be absolute to work throughout the site

lukasbestle committed:
Add missing override for reference-quicklink tpl

lukasbestle committed:
Dry main menus

distantnative commented on site/models/buzz-entry.php:
```diff
@@ -1,10 +1,11 @@
 use Kirby\Toolkit\Str;
 
-class BuzzEntryPage extends Page
+require_once __DIR__ . '/default.php';
```
`Suggestion:` Instead of having to add it to each page model, we can add this line in the `config.php` or `site` plugin's `index.php`.

distantnative commented on site/models/glossary-entry.php:
```diff
@@ -0,0 +1,11 @@
+{
+	public function menuUrl(): string
+	{
+		return $this->link()->or($this->url())->value();
```
```suggestion
		return $this->link()->toUrl() ?? $this->url();
```

For absolute URLs. And using `->toUrl()`.

distantnative commented on site/models/link.php:
```diff
@@ -0,0 +1,11 @@
+{
+	public function menuUrl(): string
+	{
+		return url($this->link()->value());
```
```suggestion
		return $this->link()->toUrl();
```

distantnative commented on site/models/reference-quicklink.php:
```diff
@@ -48,12 +49,9 @@ public static function childrenFromContentField(Field $field): array
+	public function menuUrl(): string
 	{
-		return $this->link()->toPage()?->isOpen() ?? false;
+		return url($this->link()->value());
```
```suggestion
		return $this->link()->toUrl();
```

distantnative commented on site/models/release-guide.php:
```diff
@@ -1,9 +1,14 @@
 {
+	public function menuUrl(): string
+	{
+		return $this->link()->or($this->url())->value();
```
```suggestion
		return $this->link()->toUrl() ?? $this->url();
```

distantnative commented on site/snippets/layouts/menu-items.php:
```diff
@@ -0,0 +1,9 @@
+	<?php if ($target instanceof \Kirby\Cms\Page): ?>
+	<li><a href="<?= $target->menuUrl() ?>"><?= $label ?></a></li>
+	<?php elseif (is_string($label) === true): ?>
+	<li><a <?php if ($externalClass ?? null): ?>class="<?= $externalClass ?>" <?php endif ?>href="<?= $target ?>"><?= $label ?></a></li>
```
```suggestion
	<li><a <?= e($externalClass ?? null, 'class="' . $externalClass . '" ') ?>href="<?= $target ?>"><?= $label ?></a></li>
```

distantnative commented on site/snippets/layouts/menu.php:
```diff
@@ -6,67 +6,86 @@
+							'For designers' => page('features/designers'),
+							'For content creators' => page('features/creators'),
+							'For clients & agencies' => page('features/clients'),
+							'hr',
```
```suggestion
							'-',
```

Would be more consistent with Kirby itself (e.g. Panel menu).

distantnative commented on site/snippets/layouts/menu.php:
```diff
@@ -6,67 +6,86 @@
+						'items' => [
+							'Plugins' => 'https://plugins.getkirby.com',
+							'Themes' => page('themes'),
+							'hr',
```
```suggestion
							'-',
```

distantnative commented on site/snippets/layouts/menu.php:
```diff
@@ -6,67 +6,86 @@
+							'hr',
+							'Newsletter' => page('kosmos'),
+							'Buzz' => page('buzz'),
+							'hr',
```
```suggestion
							'-',
```

distantnative commented on site/snippets/layouts/menu.php:
```diff
@@ -6,67 +6,86 @@
+						'externalClass' => 'is-external',
+						'items' => [
+							'Get together' => page('meet'),
+							'hr',
```
```suggestion
							'-',
```

distantnative commented on site/snippets/layouts/menu.php:
```diff
@@ -6,67 +6,86 @@
+							'Support forum' => 'https://forum.getkirby.com',
+							'Discord chat' => 'https://chat.getkirby.com',
+							'Community map' => 'https://community.getkirby.com',
+							'hr',
```
```suggestion
							'-',
```

lukasbestle committed:
Apply suggestions from code review

Co-authored-by: Nico Hoffmann  ෴. <me@nhoffmann.com>

lukasbestle committed:
DRY default model loading

distantnative commented on site/snippets/layouts/menu-items.php:
```diff
@@ -0,0 +1,9 @@
+	<?php if ($target instanceof \Kirby\Cms\Page): ?>
+	<li><a href="<?= $target->menuUrl() ?>"><?= $label ?></a></li>
+	<?php elseif (is_string($label) === true): ?>
+	<li><a <?= e($externalClass ?? null, 'class="' . $externalClass . '" ') ?>href="<?= $target ?>"><?= $label ?></a></li>
```
`Issue:` `$externalClass` is never passed where the snippet is used. I would actually suggest to drop that variable and always add the class `is-external` here.

```suggestion
	<li><a class="is-external" href="<?= $target ?>"><?= $label ?></a></li>
```

lukasbestle replied:
It is (in the main menu). We cannot hardcode the class as the footer menu shouldn't have it.

The issue was caused by the use of `e()` that makes the string in the second argument already evaluate even if the variable is empty or doesn't exist (unlike the `if` we had before). Fixed it with another fallback. But TBH, by that point I would prefer to go back to the original version with the `if`.

distantnative replied:
Why shouldn't the footer menu have it?

lukasbestle replied:
It probably doesn't harm, but we don't use the class for anything there. In the main menu it controls the arrow icon, the footer doesn't have that.

distantnative replied:
Yes, that's why I think it would be beneficial to just include it for the footer as well. It simplifies the logic in the `menu-item` snippet quite nicely and doesn't affect (the look of) the footer currently. But if we ever would want to differentiate those, this class would also make sense in the context of the footer.

distantnative commented on site/models/reference-quicklink.php:
```diff
@@ -48,12 +47,9 @@ public static function childrenFromContentField(Field $field): array
-	/**
-	 * Returns whether the referenced page is open
-	 */
-	public function isOpen(): bool
```
`Issue:` Something seems to have been lost here. E.g. https://getkirby.test/docs/reference/objects/filesystem/dir/is-readable - on `main` "Dir" in the "Tools" section in the Reference sidebar is highlighted. With this PR it isn't anymore. Since `->isOpen()` is still used in some parts of the sidebar, we probably cannot remove this overwritten method here.

distantnative reviewed (changes requested):
Code looks great, but ran into 2 errors.

lukasbestle committed:
Fix exception

lukasbestle committed:
Add `isOpen()` method based on `menuUrl()`

distantnative commented:
"Changes requested" shows up so harsh and negative. I'll keep using "Comments" again going forward. Just wanted to indicate that those issues need to be solved before we can merge, but the UI makes it look so negative.

lukasbestle committed:
Remove externalClass variable

lukasbestle committed:
Hardcode external class
